### ISSUE TITLE
Full Dataset CCL

### ISSUE THREAD

**Issue by william-silversmith** [seung-lab/igneous#109]
This would be mad cool. The idea would be for the output to be exactly the same as if you had run cc3d on the whole volume.

I think the way to do it would be to run first pass only and extract the union find relationships with 1 voxel overlap for every block. Save the UF mapping into a file, then combine all UF mappings. Then remap all image blocks. God this would be fun. The tricky part would be to ensure that the final labeling proceeded in order correctly from 1 to N from left to right.

**Comment by william-silversmith** [comment]
Here's a sketch that works on a small dataset in-core:

```python
import cc3d
import numpy as np
import fastremap

labels = np.load("./benchmarks/connectomics.npy")[:,:,:]

class DisjointSet:
	def __init__(self):
		self.data = {}
	def makeset(self, x):
		self.data[x] = x
		return x
	def find(self, x):
		if not x in self.data:
			return None
		i = self.data[x]
		while i != self.data[i]:
			self.data[i] = self.data[self.data[i]]
			i = self.data[i]
		return i
	def union(self, x, y):
		i = self.find(x)
		j = self.find(y)
		if i is None:
			i = self.makeset(x)
		if j is None:
			j = self.makeset(y)
		if i < j:
			self.data[j] = i
		else:
			self.data[i] = j

slices = {}
final_labels = np.zeros(labels.shape, dtype=np.uint64)

equivalences = DisjointSet()

sx, sy, sz = labels.shape

n = 0
for z in range(0, sz, 128):
	for y in range(0, sy, 128):
		for x in range(0, sx, 128):
			cutout = labels[x:x+129, y:y+129, z:z+129]
			cclabels, N = cc3d.connected_components(cutout, return_N=True, out_dtype=np.uint64)
			if cclabels.size == 0:
				continue

			cclabels += n * (128 ** 3)

			for i in range(1, N+1):
				i += n * (128 ** 3)
				equivalences.makeset(i)

			slices[(x,y,z)] = [
				cclabels[:,:,-1],
				cclabels[:,-1,:],
				cclabels[-1,:,:],
			]

			final_labels[x:x+129, y:y+129, z:z+129] = cclabels

			prev = [
				slices.get((x-128,y,z), [None] * 3)[2],
				slices.get((x,y-128,z), [None] * 3)[1],
				slices.get((x,y,z-128), [None] * 3)[0],
			]

			cur = [
				cclabels[0,:,:],
				cclabels[:,0,:],
				cclabels[:,:,0],
			]

			for prev_i, cur_i in zip(prev, cur):
				if prev_i is None:
					continue

				mapping = fastremap.inverse_component_map(cur_i, prev_i)
				for k,vals in mapping.items():
					for val in vals:
						if val != 0 and k != 0:
							equivalences.union(k, val)

			n += 1

relabel = {}
next_label = 1
for key in equivalences.data.keys():
	lbl = equivalences.find(key)
	if lbl not in relabel:
		relabel[key] = next_label
		relabel[lbl] = next_label
		next_label += 1
	else:
		relabel[key] = relabel[lbl]

resolved_final_labels = fastremap.remap(final_labels, relabel, preserve_missing_labels=True)
```

**Comment by william-silversmith** [comment]
This was released in 4.3.0
